### ISSUE TITLE
Implementing littlefs on SPI NAND: MT29FG01ABA CRC issue

### ISSUE THREAD

**Issue by marcosgabrielserrano** [littlefs-project/littlefs#1044]
I am having trouble implementing littlefs on the following device: Micron MT29F 2G01ABA. However I get the following error:

`lfs.c:2077:warn: Superblock 0x0 has become unwritable`
I traced this down to `lfs_dir_commitcrc() line: 1765`:
```C
    if (crc != crc1) {
        return LFS_ERR_CORRUPT;
    }
```
(note I tested and got these values: `crc: 19e7e821  crc1: fe9dc9eb`)
Here is my configuration file, followed by the main implementations. Is there anything I am missing here or doing wrong?
(Note: On testing reads/writes do appear to work from my limited testing as denoted by "successful read" printed etc)
(have also manually tried the spi reads/writes and know they do work)
```C
const struct lfs_config lfs_cfg = {
    .read  = MT29FG01ABA_read,
    .prog  = MT29FG01ABA_prog,
    .erase = MT29FG01ABA_erase,
    .sync  = MT29FG01ABA_sync,

    // block device configuration
    .read_size = 2048,
    .prog_size = 2048,
    .block_size = 2048 * 64,
    .block_count = 2048,
    .cache_size = 2048,
    .lookahead_size = 2048,
    .block_cycles = -1,
};

int MT29F2G01ABA_read(__attribute__((unused)) const struct lfs_config *c, lfs_block_t block, lfs_off_t off, void *buffer, lfs_size_t size)
{
    int ret;
    int bytes_read = 0;

    printf("READ  Block: %"PRIu32" Offset: %"PRIu32"  Buffer: %p  Size: %"PRIu32"\n", block, off, buffer, size);

    if(block >= FLASH_BLOCK_CNT)
    {
        return -1;
    }
    if(off % FLASH_PAGE_SIZE)
    {
        return -1;
    }
    if(size % FLASH_PAGE_SIZE)
    {
        return -1;
    }

    while(bytes_read < (int)size)
    {
        PAGE_READ(get_real_address(block, off + (lfs_off_t)bytes_read));

        sleep_us(5);
        ret = verify_page_read_success(NULL);
        if(ret < 0)
        {
            printf("Failed to read\n");
            return ret;
        }

        bytes_read += READ_FROM_CACHE_1(0, (u8 *)buffer, FLASH_PAGE_SIZE);
    }

    printf("Successful Read\n");

    return 0;
}

int MT29F2G01ABA_prog(__attribute__((unused)) const struct lfs_config *c, lfs_block_t block, lfs_off_t off, const void *buffer, lfs_size_t size)
{
    int bytes_written = 0;
    int ret = 0;
    status_reg_t feature;
    
    printf("PROG: Block: %"PRIu32" Offset: %"PRIu32"  Buffer: %p  Size: %"PRIu32"\n", block, off, buffer, size);


    if(block >= FLASH_BLOCK_CNT)
    {
        printf("Block value out of bounds\n");
        return -1;
    }
    if(off % FLASH_PAGE_SIZE)
    {
        printf("Page selection not aligned\n");
        return -1;
    }
    if(off >= FLASH_PAGE_CNT)
    {
        printf("Offset value out of bounds\n");
        return -1;
    }
    if(size % FLASH_PAGE_SIZE)
    {
        printf("Writing incorrect amount of bytes\n");
        return -1;
    }

    while(bytes_written < (int)size)
    {

        WRITE_ENABLE();

        if(wait_write_enabled() < 0) // is this even necessary?
        {
            // need to notify error here somehow
            printf("PROG: Write failed to enable\n");
            return -1;
        }
        // prog_load
        // in all actuality, I may have to reduce this size to like 512 or less due to limited stack space
        bytes_written += PROGRAM_LOAD_1(0, (u8 *)(buffer + bytes_written), FLASH_PAGE_SIZE); // if prog size changed, this needs to change too
        
        PROGRAM_EXECUTE(get_real_address(block, off + (lfs_off_t)bytes_written));

        // check status via get_feature
        sleep_us(5);
        ret = verify_program_execute_success(&feature); // set success/timeout errorcode that bubbles up
        if(ret < 0)
        {
            printf("PROG: Write failed to verify execute success\n");
            // do I reset previous writes here or something?
            return ret;
        }
    }
    printf("Successful Program!  Feature reg: %"PRIx8"\n", feature.reg_value);

    return ret;
}

int MT29F2G01ABA_erase(__attribute__((unused)) const struct lfs_config *c, lfs_block_t block)
{
    status_reg_t feature;

    printf("Erasing\n");

    WRITE_ENABLE();

    BLOCK_ERASE(get_real_address(block, 0));

    return verify_block_erase_success(&feature);
}

int MT29F2G01ABA_sync(__attribute__((unused)) const struct lfs_config *c)
{
    printf("Syncing\n");
    return 0;
}
```

**Comment by marcosgabrielserrano** [comment]
Update:

Strange, found that right before this, there is a write of the dir commit via PROG to the chip, then it is read to read the hash.

However though I See it writing some string of bytes corresponding to dir commit, when it reads it is reading only a bunch of 0xFFs with same input block/offsets etc. 
(Confirmed with other code that prog + subsequent read works correctly so I don't understand why'd it'd fail here)

**Comment by marcosgabrielserrano** [comment]
Update 2:

I've investigated further by playing with the READ and PROG functions, and I've found that most likely, there's some sort of problem with either 1. the NAND chip, 2. how I use it, 3. power??(cause i don't know what else?)

After testing the functions themselves without littlefs I ran into this odd behavior:

1. Enable Write
2. Load 2048 bytes to NAND flash
3. Program Execute to write to Block 0 (or any block)
4. Enable Write
5. Load 2048 bytes to NAND flash (different bytes to see diff)
6. Program Execute to Block 1
7. Read Block 0 -> Read successfully retrieves written bytes
8. Read Block 1 -> Fails -> 0xFF returned (default unwritten byte value)

For some reason the first block that is written successfully seems to "store" the data and read it back but the 2nd and after don't.

And to further confuse the matter, ALL the data is all lost after a power cycle. Which is strange as isn't this data supposed to be retained? (it's not RAM after all - unless someone wants to correct me?)

So current theory is the problem is not littlefs, but rather something with the nand chip/implementation... though I can't say I know what as it matches the documenation/online examples I see in terms of order etc...
